libarchive/test/test_read_format_rar_encryption.c

```diff
@@ -0,0 +1,176 @@
+		assertEqualInt(ARCHIVE_FATAL, archive_read_data(a, buff, sizeof(buff)));
+
+		/* Additional attempts to read headers are futile */
+		assertEqualInt(ARCHIVE_OK, archive_read_close(a));
+		assertEqualInt(ARCHIVE_OK, archive_read_free(a));
+		return;
+	}
+
+	/* From here on out, we can assume that !filenamesEncrypted */
+
+	/* Read the header for "b.txt" */
+	assertEqualIntA(a, ARCHIVE_OK, archive_read_next_header(a, &ae));
+	assertEqualInt((AE_IFREG | 0644), archive_entry_mode(ae));
+	assertEqualString("b.txt", archive_entry_pathname(ae));
+	assertEqualInt(expected_file_size, archive_entry_size(ae));
+	assertEqualInt(1, archive_entry_is_data_encrypted(ae));
+	assertEqualInt(0, archive_entry_is_metadata_encrypted(ae));
+	assertEqualInt(1, archive_read_has_encrypted_entries(a));
+	assertEqualIntA(a, ARCHIVE_FATAL, archive_read_data(a, buff, sizeof(buff)));
+
+	/* Read the header for "c.txt" */
+	assertEqualIntA(a, ARCHIVE_OK, archive_read_next_header(a, &ae));
+	assertEqualInt((AE_IFREG | 0644), archive_entry_mode(ae));
+	assertEqualString("c.txt", archive_entry_pathname(ae));
+	assertEqualInt(expected_file_size, archive_entry_size(ae));
+	assertEqualInt(solid, archive_entry_is_data_encrypted(ae));
+	assertEqualInt(0, archive_entry_is_metadata_encrypted(ae));
+	/* After setting to true above, this should forever be true */
+	assertEqualInt(1, archive_read_has_encrypted_entries(a));
+	assertEqualIntA(a, expected_read_data_result, archive_read_data(a, buff, sizeof(buff)));
+	if (!solid) {
+		assertEqualMem("This is from c.txt", buff, expected_file_size);
+	}
+
+	/* Read the header for "d.txt" */
+	assertEqualIntA(a, ARCHIVE_OK, archive_read_next_header(a, &ae));
+	assertEqualInt((AE_IFREG | 0644), archive_entry_mode(ae));
+	assertEqualString("d.txt", archive_entry_pathname(ae));
+	assertEqualInt(expected_file_size, archive_entry_size(ae));
+	assertEqualInt(1, archive_entry_is_data_encrypted(ae));
+	assertEqualInt(0, archive_entry_is_metadata_encrypted(ae));
+	assertEqualInt(1, archive_read_has_encrypted_entries(a));
+	assertEqualIntA(a, ARCHIVE_FATAL, archive_read_data(a, buff, sizeof(buff)));
+
+	/* End of archive. */
+	assertEqualIntA(a, ARCHIVE_EOF, archive_read_next_header(a, &ae));
+
+	/* Close the archive. */
+	assertEqualInt(ARCHIVE_OK, archive_read_close(a));
+	assertEqualInt(ARCHIVE_OK, archive_read_free(a));
+}
+
+DEFINE_TEST(test_read_format_rar4_encrypted)
+{
+	test_encrypted_rar_archive("test_read_format_rar4_encrypted.rar", 0, 0);
+}
+
+DEFINE_TEST(test_read_format_rar4_encrypted_filenames)
+{
+	test_encrypted_rar_archive("test_read_format_rar4_encrypted_filenames.rar", 1, 0);
+}
+
+DEFINE_TEST(test_read_format_rar4_solid_encrypted)
+{
+	/* TODO: If solid RAR4 support is ever added, the following should pass */
+#if 0
+	test_encrypted_rar_archive("test_read_format_rar4_solid_encrypted.rar", 0, 1);
+#else
+	skipping("RAR4 solid archive support not currently available");
+#endif
+}

```

Comment:
I figure that at least having this here will make someone's life easier if RAR4 solid support is ever added. I can remove if folks don't want currently dead code laying around.